### ISSUE TITLE
Unsupported grant type when trying to get the token

### ISSUE THREAD

**Issue by g-alvarez** [jazzband/django-oauth-toolkit#479]
Hello, I installed DOT and i made an application with "Authorization grant type" set as  "Resource owner password-based" and "client-type" set as "Public". Using Postman i try to send a POST request to the /o/token/ url with the following:

- Header
Content-Type: application/x-www-form-urlencoded

- Body
client_id: "MY_CLIENT_ID"
grant_type: password
username: "MY_USERNAME"
password: "MY_PASSWORD"

and when i send the request i get this response: 

{
    "error": "unsupported_grant_type"
}

This was all done using django 1.10 and python 2.7. I'm not sure what exactly i'm doing wrong, can you guys help?. Thanks in advance for the reply.

**Comment by aniruddha-adhikary** [comment]
client-type should be private. You should send client id and secret as basic http auth username and password.

**Comment by angelfeliz** [comment]
I have the same problem

**Comment by pirm-in** [comment]
jesus.. this took me ages to figure out. for documentation: find the answer here https://github.com/evonove/django-oauth-toolkit/issues/127

in essence:
problem is that with axios requests the submitted data is not in the request's POST field, but in the body. you can change the behaviour of DOT with a setting, so that it reads the submitted data from there.

OAUTH2_PROVIDER = {
    # other OAUTH2 settings
    'OAUTH2_BACKEND_CLASS': 'oauth2_provider.oauth2_backends.JSONOAuthLibCore'
}

**Comment by sajib49** [comment]
https://stackoverflow.com/questions/47795356/reddit-api-unsupported-grant-type-error-with-retrofit-java

**Comment by florianmartens** [comment]
The answer from @prim-in was right and helped me. However, the explanation may be a little cryptic. 

The the problem is that the django oauth toolkit expects your frontend to use this header for content-type:
`"Content-Type": "application/x-www-form-urlencoded"`

When you are using a JavaScript/JSON frontend you are, however, quite likely to send a Content-Type header with "application/json". If you use the "application/json" header, DOT can't read the request.POST object in the django backend. So the object is empty and your grant_type will not been seen by it.

You have to tell DOT that you are using JSON to send the data. You do this by going into your django settings.py file and adding/changing this:

```
OAUTH2_PROVIDER = {
# other OAUTH2 settings
'OAUTH2_BACKEND_CLASS': 'oauth2_provider.oauth2_backends.JSONOAuthLibCore'
}

```

It is explained quite will in the link @pirm-in prim-in provided

**Comment by Blanen** [comment]
I get this error on the literal rest example... Not great.

**Comment by jordotech** [comment]
> I get this error on the literal rest example... Not great.

@Blanen I'm running into the same problems as you, very frustrated with the documentation.  Please let us know if you figured it out.

**Comment by shellbot97** [comment]
> > I get this error on the literal rest example... Not great.
> 
> @Blanen I'm running into the same problems as you, very frustrated with the documentation. Please let us know if you figured it out.

were you able to figure this out, finally?!

**Comment by Blanen** [comment]
@shellbot97 no

**Comment by MorozovVasily** [comment]
> > > I get this error on the literal rest example... Not great.
> > 
> > 
> > @Blanen I'm running into the same problems as you, very frustrated with the documentation. Please let us know if you figured it out.
> 
> were you able to figure this out, finally?!

I figure this out!
As I can see, this problem appear 12 days ago, so I look at [previous version](https://web.archive.org/web/20191225132059/https://django-oauth-toolkit.readthedocs.io/en/latest/rest-framework/getting_started.html) of rest example: 
And find difference
if **don't** write: OAUTH2_BACKEND_CLASS': 'oauth2_provider.oauth2_backends.JSONOAuthLibCore',
[this](https://django-oauth-toolkit.readthedocs.io/en/latest/rest-framework/getting_started.html#step-4-get-your-token-and-use-your-api) rest example will work

**Comment by n2ygk** [comment]
Yes, with apologies, we merged a documentation error. See https://github.com/jazzband/django-oauth-toolkit/issues/807#issuecomment-601691819.

I don't see evidence in a quick scan of the OAuth 2.0 [RFC 6749](https://tools.ietf.org/html/rfc6749) that one is allowed to POST with a JSON body rather than form-encoded.... If you see a reference otherwise, please let me know and/or submit a PR.

I've used https://github.com/manfredsteyer/angular-oauth2-oidc and it does the POSTs using form-encoding.

**Comment by Jogendra1234** [comment]
Hello,
I am getting the same error, please figure out this and let me know it's solution.
error":"unsupported_grant_type"

**Comment by alovega** [comment]
> > I get this error on the literal rest example... Not great.
> 
> @Blanen I'm running into the same problems as you, very frustrated with the documentation. Please let us know if you figured it out.
I really can't make sense of the documentation, I am trying to use it to authorize a django user

**Comment by Stolek007** [comment]
Set header -> "Content-Type" => "application/json"

**Comment by nicbou** [comment]
I also get the same error. It would be so much easier to solve this issue if we got a useful error message in the response body, or if this was better covered in the documentation.

## tl;dr

Set an appropriate Content-Type header for your token request. `application/x-www-form-urlencoded` works out of the box.

My POST request looks like this (vanilla JS):

```

  async getAuthToken(authorizationCode) {
    const tokenUrl = new URL(`oauth/token/`, this.getApiBase());
    const bodyParams = new URLSearchParams({
      grant_type: 'authorization_code',
      code: authorizationCode,
      client_id: 'uiAJcCMKHy8XqyIfEz88tnu0o2osyYl2owSTNaUa',
      code_verifier: this.codeVerifier,
      redirect_uri: `https://${window.location.hostname}/oauth-redirect`,
    });

    return fetch(tokenUrl, {
      method: 'POST',
      headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
      body: bodyParams.toString(),
    });
  }
```

**Comment by Izocel** [comment]
![image](https://github.com/jazzband/django-oauth-toolkit/assets/68454661/621611ea-6ae0-4aee-8eeb-ff9c44b81651)
